src/awkward/_v2/_connect/rdataframe/to_rdataframe.py

```diff
@@ -225,19 +242,21 @@ class {array_data_source} final
         }}
     }};
 
-    ROOT::RDataFrame* MakeAwkwardArrayDS_{array_data_source}(ULong64_t size, std::initializer_list<ULong64_t> ptrs_list) {{
-        return new ROOT::RDataFrame(std::make_unique<{array_data_source}>(size, ptrs_list));
+    ROOT::RDataFrame* MakeAwkwardArrayDS_{array_data_source}(PyObject* lookup, ULong64_t size, std::initializer_list<ULong64_t> ptrs_list) {{
+        return new ROOT::RDataFrame(std::make_unique<{array_data_source}>(std::forward<PyObject*>(lookup), size, ptrs_list));
     }}
 
 }}
             """
 
-            done = compiler(cpp_code)
+            done = compile(cpp_code)
             assert done is True
 
         rdf = getattr(ROOT.awkward, f"MakeAwkwardArrayDS_{array_data_source}")(
+            self.lookups,
             self.length,
             (self.data_ptrs_list),
         )
```

Comment:
Does ROOT change the reference count when you return a PyObject? (And thus, we'd have to be careful not to overcount...)

```python
>>> import ROOT
>>> ROOT.gInterpreter.Declare("""
... #include <Python.h>
... 
... PyObject* g(PyObject* obj) {
...     std::cout << "HERE " << Py_REFCNT(obj) << std::endl;
...     return obj;
... }
... """)
True
>>> import sys
>>> some_object = (1, 2.2, "three")
>>> sys.getrefcount(some_object)
2
>>> another_reference = ROOT.g(some_object)
HERE 3
>>> sys.getrefcount(some_object)
2
>>> del some_object
>>> anoth *** Break *** segmentation violation
```

Nope!

So the safe way to do this is to manually increment:

```python
>>> import ROOT
>>> ROOT.gInterpreter.Declare("""
... #include <Python.h>
... 
... PyObject* g(PyObject* obj) {
...     std::cout << "HERE " << Py_REFCNT(obj) << std::endl;
...     Py_INCREF(obj);
...     return obj;
... }
... """)
True
>>> import sys
>>> some_object = (1, 2.2, "three")
>>> sys.getrefcount(some_object)
2
>>> another_reference = ROOT.g(some_object)
HERE 3
>>> sys.getrefcount(some_object)
3
>>> del some_object
>>> another_reference
(1, 2.2, 'three')
```

That means that we could implement an accessor on our RDataSource by writing a method like

```c++
PyObject* GetLookup() {
    Py_INCREF(fPyLookup);
    return fPyLookup;
}
```

possibly with GIL-safety, but probably not (because it would only be called from Python, and presumably PyROOT doesn't release the GIL for a direct call like this).

Then we would just need to get from an RDataFrame object to the RDataSource at the root of its DAG.